packages/flutter/lib/src/material/slider_parts.dart

```diff
@@ -830,6 +830,66 @@ class DropSliderValueIndicatorShape extends SliderComponentShape {
       sizeWithOverflow: sizeWithOverflow,
       backgroundPaintColor: sliderTheme.valueIndicatorColor!,
       strokePaintColor: sliderTheme.valueIndicatorStrokeColor,
+      sliderTheme: sliderTheme,
+    );
+  }
+}
+
+/// Utility class for processing multiline text in value indicators.
+abstract class _ValueIndicatorTextProcessor {
+  const _ValueIndicatorTextProcessor._();
+
+  /// Creates a TextPainter with ellipsis when text exceeds maxLines.
+  static TextPainter _createTextPainterWithEllipsis({
+    required TextPainter originalPainter,
+    required String originalText,
+    required int maxLines,
+    required double textScaleFactor,
+  }) {
+    final List<String> lines = originalText.split('\n');
+    final List<String> visibleLines = lines.take(maxLines - 1).toList();
+    visibleLines.add('${lines[maxLines - 1]}...');
+
+    return TextPainter(
+      text: TextSpan(text: visibleLines.join('\n'), style: originalPainter.text?.style),
+      textDirection: originalPainter.textDirection,
+      textScaleFactor: textScaleFactor,
+    )..layout();
+  }
+
+  /// Processes multiline text for value indicators, handling maxLines
+  /// truncation and ellipsis.
+  ///
+  /// Returns the appropriate TextPainter based on the multiline configuration.
+  /// The caller is responsible for disposing the returned TextPainter
+  /// if it differs from the original.
+  static TextPainter processMultilineText({
+    required TextPainter originalPainter,
+    required ValueIndicatorMultilineConfig multilineConfig,
+    required double textScaleFactor,
+  }) {
+    if (!multilineConfig.enabled || multilineConfig.maxLines == null) {
+      return originalPainter;
+    }
+
+    final TextPainter truncatedPainter = TextPainter(
+      text: originalPainter.text,
+      textDirection: originalPainter.textDirection,
+      textScaleFactor: textScaleFactor,
+      maxLines: multilineConfig.maxLines,
+    )..layout();
+
+    if (!truncatedPainter.didExceedMaxLines) {
```

Comment:
Looks like the text is laid out in an infinite canvas. IIRC there's a numberOfLines field on text painter (or ui.Paragraph) that you can use to figure out the number of visible lines. Also `RenderEditable` may have logic for calculating the number of hard line breaks in a paragraph. Maybe you can avoid creating new text painters by replicating that logic or by using `numberOfLines`?